### PR TITLE
Get EmptyHttpResult in RDF via reflection

### PR DESCRIPTION
**Pull request by captainsafia**
Addresses https://github.com/dotnet/aspnetcore/issues/45063

### CODE REVIEW

File: src/Http/Http.Extensions/src/RequestDelegateFactory.cs [hunk 1]
```diff
@@ -92,7 +92,10 @@ public static partial class RequestDelegateFactory
     private static readonly MemberExpression FormFilesExpr = Expression.Property(FormExpr, typeof(IFormCollection).GetProperty(nameof(IFormCollection.Files))!);
     private static readonly MemberExpression StatusCodeExpr = Expression.Property(HttpResponseExpr, typeof(HttpResponse).GetProperty(nameof(HttpResponse.StatusCode))!);
     private static readonly MemberExpression CompletedTaskExpr = Expression.Property(null, (PropertyInfo)GetMemberInfo<Func<Task>>(() => Task.CompletedTask));
-    private static readonly NewExpression EmptyHttpResultValueTaskExpr = Expression.New(typeof(ValueTask<object>).GetConstructor(new[] { typeof(EmptyHttpResult) })!, Expression.Property(null, typeof(EmptyHttpResult), nameof(EmptyHttpResult.Instance)));
```
Comment (davidfowl) [inline src/Http/Http.Extensions/src/RequestDelegateFactory.cs:95]: Delete the EmptyHttpResult type

File: src/Http/Http.Extensions/test/RequestDelegateFactoryTests.cs [hunk 2]
```diff
@@ -7050,6 +7062,33 @@ public void Create_Populates_EndpointBuilderWithRequestDelegateAndMetadata()
...
+        var @delegate = (string name, HttpContext context) => context.Items.Add("param", name);
+
+        var result = RequestDelegateFactory.Create(@delegate, new RequestDelegateFactoryOptions()
+        {
+            EndpointBuilder = CreateEndpointBuilderFromFilterFactories(new List<Func<EndpointFilterFactoryContext, EndpointFilterDelegate, EndpointFilterDelegate>>()
+            {
+                (routeHandlerContext, next) => async (context) =>
+                {
+                    var response = await next(context);
+                    Assert.IsType<EmptyHttpResult>(response);
```
Comment (davidfowl) [inline src/Http/Http.Extensions/test/RequestDelegateFactoryTests.cs:7077]:
```suggestion
                    Assert.IsType<EmptyHttpResult>(response);
                    Assert.Same(Results.Empty, response);
```

File: src/Http/Http.Extensions/src/RequestDelegateFactory.cs [hunk 3]
```diff
@@ -92,8 +92,11 @@ public static partial class RequestDelegateFactory
...
-
+    // Due to https://github.com/dotnet/aspnetcore/issues/41330 we cannot reference the EmptyHttpResult type
+    // but users still need to assert on it as in https://github.com/dotnet/aspnetcore/issues/45063
+    // so we temporarily work around this here by using reflection to get the actual type.
+    private static readonly NewExpression EmptyHttpResultValueTaskExpr = Expression.New(typeof(ValueTask<object>).GetConstructor(new[] { typeof(IResult) })!, Expression.Property(null, Type.GetType("Microsoft.AspNetCore.Http.Results, Microsoft.AspNetCore.Http.Results")!.GetProperty("Empty")!));
+    private static readonly object? EmptyHttpResultInstance = Type.GetType("Microsoft.AspNetCore.Http.HttpResults.EmptyHttpResult, Microsoft.AspNetCore.Http.Results")!.GetProperty("Instance")!.GetValue(null, null);
```
Comment (mitchdenny) [inline src/Http/Http.Extensions/src/RequestDelegateFactory.cs:99]: More of a question than a blocking comment. Can AOT navigate this use of reflection?

Comment (davidfowl) [reply src/Http/Http.Extensions/src/RequestDelegateFactory.cs:99]: Nope, this is fully AOT unsafe. This code path isn't fully AOT safe to begin with (see https://github.com/dotnet/aspnetcore/pull/42272).

**Pull request closed without merging**